File: src/main/scala/gay/menkissing/spectrumstorage/util/FluidResource.scala
package gay.menkissing.spectrumstorage.util

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.component.DataComponentPatch
import net.minecraft.core.registries.{BuiltInRegistries, Registries}
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.{ByteBufCodecs, StreamCodec}
import net.minecraft.world.level.material.{FlowingFluid, Fluid, Fluids}
import net.neoforged.neoforge.fluids.{FluidStack, FluidType}

import java.util.Objects

final class FluidResource private (val fluid: Fluid, val components: DataComponentPatch) {
  override val hashCode: Int = Objects.hash(fluid, components)

  def isBlank: Boolean =
    fluid == Fluids.EMPTY

  def componentsMatch(other: DataComponentPatch): Boolean =
    this.components == other

  def makeStack(size: Int): FluidStack =
    if size == 0 then
      FluidStack.EMPTY
    else
      FluidStack(fluid.builtInRegistryHolder(), size, components)

  def sameAsStack(stack: FluidStack): Boolean =
    this.fluid == stack.getFluid && this.componentsMatch(stack.getComponentsPatch)
  
  override def equals(obj: Any): Boolean =
    if (this eq obj.asInstanceOf[Object]) true
    else if (obj == null || getClass != obj.getClass) false
    else
      val that = obj.asInstanceOf[FluidResource]
      hashCode == that.hashCode && fluid == that.fluid && componentsMatch(that.components)
}

object FluidResource:
  
  val CODEC: Codec[FluidResource] =
    RecordCodecBuilder.create: builder =>
      builder.group(
        BuiltInRegistries.FLUID.byNameCodec().fieldOf("fluid").forGetter((it: FluidResource) => it.fluid),
        DataComponentPatch.CODEC.fieldOf("components").forGetter((it: FluidResource) => it.components)
      ).apply(builder, FluidResource.apply)
  val STREAM_CODEC: StreamCodec[RegistryFriendlyByteBuf, FluidResource] =
    StreamCodec.composite(
      ByteBufCodecs.registry(Registries.FLUID), (it: FluidResource) => it.fluid,
      DataComponentPatch.STREAM_CODEC, (it: FluidResource) => it.components,
      FluidResource.of
    )



  val EMPTY = FluidResource(Fluids.EMPTY, DataComponentPatch.EMPTY)

  def ofStack(stack: FluidStack): FluidResource =
    Objects.requireNonNull(stack, "Fluid stack may not be null")

    FluidResource(stack.getFluid, stack.getComponentsPatch)

  def of(fluid: Fluid): FluidResource =
    of(fluid, DataComponentPatch.EMPTY)
  
  def of(fluid: Fluid, components: DataComponentPatch): FluidResource =
    Objects.requireNonNull(fluid, "Fluid may not be null")
    Objects.requireNonNull(components, "Components may not be null")

    val goodFluid =
      if (!fluid.isSource(fluid.defaultFluidState()) && fluid != Fluids.EMPTY) {
        fluid match
          case it: FlowingFluid => it.getSource()
          case _ =>
            val id = BuiltInRegistries.FLUID.getKey(fluid)
            throw new IllegalArgumentException(s"Can't convert flowing fluid $id ($fluid) into a still fluid")
      } else fluid

    FluidResource(goodFluid, components)

File: src/main/java/gay/menkissing/pastelstorage/api/fluid/FluidResource.java
package gay.menkissing.pastelstorage.api.fluid;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.component.DataComponentPatch;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.level.material.FlowingFluid;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.world.level.material.Fluids;
import net.neoforged.neoforge.fluids.FluidStack;

import java.util.Locale;
import java.util.Objects;

public final class FluidResource {
    private final Fluid fluid;
    private final DataComponentPatch components;
    private final int hashCode;

    private FluidResource(Fluid fluid, DataComponentPatch components) {
        this.hashCode = Objects.hash(fluid, components);
        this.fluid = fluid;
        this.components = components;
    }

    public boolean isBlank() {
        return fluid.isSame(Fluids.EMPTY);
    }

    public FluidStack makeStack(int size) {
        return size == 0 ? FluidStack.EMPTY : new FluidStack(fluid.builtInRegistryHolder(), size, components);
    }

    public boolean componentsMatch(DataComponentPatch other) {
        return this.components.equals(other);
    }

    public boolean sameAsStack(FluidStack stack) {
        return this.fluid.equals(stack.getFluid()) && this.componentsMatch(stack.getComponentsPatch());
    }

    public Fluid getFluid() {
        return fluid;
    }

    public DataComponentPatch getComponents() {
        return components;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null || !getClass().equals(obj.getClass())) {
            return false;
        } else {
            FluidResource that = (FluidResource)obj;
            return this.hashCode == that.hashCode && this.fluid.equals(that.fluid) && componentsMatch(that.components);
        }
    }


    @Override
    public int hashCode() {
        return this.hashCode;
    }

    public static final Codec<FluidResource> CODEC =
            RecordCodecBuilder.create(builder ->
                builder.group(
                    BuiltInRegistries.FLUID.byNameCodec().fieldOf("fluid").forGetter(FluidResource::getFluid),
                    DataComponentPatch.CODEC.fieldOf("components").forGetter(FluidResource::getComponents)
                ).apply(builder, FluidResource::new)
            );
    
    public static final StreamCodec<RegistryFriendlyByteBuf, FluidResource> STREAM_CODEC =
            StreamCodec.composite(
                    ByteBufCodecs.registry(Registries.FLUID), FluidResource::getFluid,
                    DataComponentPatch.STREAM_CODEC, FluidResource::getComponents,
                    FluidResource::of
            );
            
    public static final FluidResource EMPTY = new FluidResource(Fluids.EMPTY, DataComponentPatch.EMPTY);
    
    public static FluidResource ofStack(FluidStack stack) {
        Objects.requireNonNull(stack, "Fluid stack may not be null");

        return new FluidResource(stack.getFluid(), stack.getComponentsPatch());
    }

    public static FluidResource of(Fluid fluid) {
        return of(fluid, DataComponentPatch.EMPTY);
    }
    
    public static FluidResource of(Fluid fluid, DataComponentPatch components) {
        Objects.requireNonNull(fluid, "Fluid may not be null");
        Objects.requireNonNull(components, "Components may not be null");

        var goodFluid = fluid;

        if (!fluid.isSource(fluid.defaultFluidState()) && !fluid.equals(Fluids.EMPTY)) {
            if (fluid instanceof FlowingFluid flowingFluid) {
                goodFluid = flowingFluid.getSource();
            } else {
                var id = BuiltInRegistries.FLUID.getKey(fluid);
                throw new IllegalArgumentException(String.format(Locale.ROOT, "Can't convert flowing fluid %s (%s) into a still fluid", id.toString(), fluid.toString()));
            }
        }
        
        return new FluidResource(goodFluid, components);
    }

}
